### PR TITLE
[SPARK-10154][SQL] remove the no-longer-necessary CatalystScan

### PR DESCRIPTION
**Pull request by cloud-fan**
There is no build-in data source depending on it and no example to demonstrate how to use it.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala [hunk 1]
```diff
@@ -278,44 +301,6 @@ trait PrunedFilteredScan {
...
- * 3. It assumes that fields of the data provided in the insert method are nullable.
- * If a data source needs to check the actual nullability of a field, it needs to do it in the
- * insert method.
- *
- * @since 1.3.0
- */
-@DeveloperApi
-trait InsertableRelation {
-  def insert(data: DataFrame, overwrite: Boolean): Unit
-}
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: I moved it to the relation section.

**Comment by cloud-fan** [comment]
cc @liancheng

**Comment by liancheng** [comment]
LGTM, please fix the MiMa failure by adding

```scala
ProblemFilters.exclude[MissingClassProblem]("org.apache.spark.sql.sources.CatalystScan")
```

to `project/MimaExcludes.scala`.

**Comment by liancheng** [comment]
Hey @cloud-fan, [Michael's comment] [1] reminds me that I had actually discussed once with Michael whether we should remove `CatalystScan`, and the conclusion was we can keep it there as an  experimental interface. It would be convenient to experiment with new data sources since Catalyst expression classes are more flexible.

[1]: https://github.com/apache/spark/pull/3269/files#r37668626

**Comment by cloud-fan** [comment]
ah, ok let's keep it :)

**Pull request closed without merging**